RPC query account information and access keys by a specific block id

frol opened the pull request:
*Resolves #2015*

bowenwang1996 commented on chain/jsonrpc/src/lib.rs:
```diff
@@ -267,16 +269,63 @@ impl JsonRpcHandler {
+                            .map_err(|_| RpcError::server_error(Some("Invalid public key")))?,
+                    },
+                    _ => {
+                        unreachable!("path parts are at least 2 and at most 3 elements due to splitn(3) and the check right after it");
```
maybe mention that this restriction is for `access_key` query?

bowenwang1996 commented on core/primitives/src/views.rs:
```diff
@@ -165,7 +165,16 @@ pub enum QueryResponseKind {
     Error(QueryError),
     AccessKey(AccessKeyView),
     AccessKeyList(AccessKeyList),
-    Validators(EpochValidatorInfo),
```
Why is this deleted?

frol replied:
This variant is unused :man_shrugging:

frol commented:
@nearmax @bowenwang1996 This PR is ready to be merged.

frol commented:
This PR introduced a new extended API which is backward-compatible (the old API of `["account/test.near", ""]` is preserved):

```
$ http post http://127.0.0.1:3030/ jsonrpc=2.0 id=dontcare \
    method=query \
    'params:={"request_type": "view_access_key_list", "account_id": "test.near", "block_id": 300 }'
```

We are heading away from the positional-based APIs for easier maintainability. For example, it will make adding `is_final`-like extensions (https://github.com/nearprotocol/nearcore/pull/1890/) easier and more natural.

nearmax commented on chain/client/src/view_client.rs:
```diff
@@ -109,17 +110,26 @@ impl ViewClientActor {
+        };
+        let header = header.map_err(|e| e.to_string())?.clone();
+
+        let account_id = match &msg.request {
```
I think it is a good call to not extract `account_id` to `RpcQueryRequest`

nearmax commented:
Let's have a documentation issue for it :)

frol commented:
@evgenykuzyakov GitHub says this PR requires your review before we merge

frol commented:
Oh, I forgot to make the base64-encoding for the `Vec<u8>` fields. See #2034.

Let me make this change before we merge.

frol commented:
I have fixed the API to handle base64-encoded strings instead of crazy arrays of numbers. To do that, I introduced two new types: `StoreKey` (I plan to introduce it throughout the code base down to `core/store/` in #2034) and `FunctionArgs` (I will create a separate PR to finish the migration to this type in the rest code base) to replace ad-hoc `Vec<u8>`, and to be able to do automatic base64 encoding with serde. I am planning to introduce even more named wrapping types to replace `Vec<u8>`.

@nearmax @bowenwang1996 @evgenykuzyakov Please, review.

evgenykuzyakov commented on chain/chain/src/test_utils.rs:
```diff
@@ -665,31 +666,27 @@ impl RuntimeAdapter for KeyValueRuntime {
+            QueryRequest::ViewAccount { account_id, .. } => Ok(QueryResponse {
+                kind: QueryResponseKind::ViewAccount(
+                    Account {
+                        amount: self.state.read().unwrap().get(&state_root).map_or_else(
```
map_or?

frol replied:
Well, that might be undesirable. Anyway, I have not touched this code, it is only the indentation that is changed in this line. We will need to address the `unwrap`s in a security review round.

evgenykuzyakov commented on chain/client/src/view_client.rs:
```diff
@@ -109,17 +110,26 @@ impl ViewClientActor {
+            QueryRequest::ViewAccessKeyList { account_id, .. } => account_id,
+            QueryRequest::CallFunction { account_id, .. } => account_id,
+        };
+        let shard_id = self.runtime_adapter.account_id_to_shard_id(account_id);
```
Do you need to check account ID validity before converting it to shard here?

frol replied:
Well, I would like to improve `AccountId` type, so it enforces the contract of being a valid account ID. Let's make it in another PR; created an issue for that: https://github.com/nearprotocol/nearcore/issues/2074

evgenykuzyakov commented on core/primitives/src/serialize.rs:
```diff
@@ -76,6 +78,34 @@ pub trait BaseDecode: for<'a> TryFrom<&'a [u8], Error = Box<dyn std::error::Erro
     }
 }
 
+macro_rules! pretty_bytes_serialization {
```
Why is it called `pretty`? Should it be `base64_bytes_serialization`?

frol replied:
Well, this is a reasonable comment, though, my thinking around this is that base64 is just an implementation detail, and the intent is to be able to serialize bytes into a "pretty format".
